### PR TITLE
Database and new architecture

### PR DESCRIPTION
**Pull request by danilkolikov**
Database fixed for new architecture
Added ignored tests

### CODE REVIEW

File: app/src/androidTest/java/ly/loud/loudly/util/database/AttachmentTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import ly.loud.loudly.new_base.LoudlyImage;
+import ly.loud.loudly.new_base.interfaces.attachments.Attachment;
+import ly.loud.loudly.util.Equality;
+import ly.loud.loudly.test.Generators;
+import org.junit.Assert;
+
+/**
+ * @author Danil Kolikov
+ */
+public class AttachmentTest extends DatabaseTest<ly.loud.loudly.new_base.interfaces
```
Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/AttachmentTest.java]: Is there another attachment?

File: app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+import static org.junit.Assert.*;
+
+/**
+ * @author Danil Kolikov
+ */
+@RunWith(AndroidJUnit4.class)
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public abstract class DatabaseTest<T> {
+    protected Random random;
```
Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java]: nullability

File: app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    protected StorIOSQLite postsDatabase;
```
Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java]: same

File: app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    public static void cleanTables(@NonNull StorIOSQLite database, String... tables) {
```
Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java]: same

File: app/src/androidTest/java/ly/loud/loudly/util/database/AttachmentTest.java (hunk 1 above)
Comment (danilkolikov) [inline app/src/androidTest/java/ly/loud/loudly/util/database/AttachmentTest.java]: Yes, there is the Attachment in util.database.entities

File: app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java (hunk 4 above)
Comment (danilkolikov) [inline app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java]: :(

File: app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    @Test
+    public void test_01_one() throws DatabaseException {
+        T object = generate();
+        long id = insert(object);
+        checkStored(id, object);
+
+        delete(id);
+        checkAbsent(id);
```
Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java]: Maybe we need Mockito for such kind of checks?

File: app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+ * @author Danil Kolikov
+ */
+@RunWith(AndroidJUnit4.class)
+public class KeyTest {
+    private StorIOSQLite keysDatabase;
+    private Random random;
+
+    private static void checkKeys(@NonNull List<KeyKeeper> keys) {
+        for (int i = 0; i < Networks.NETWORK_COUNT; i++) {
+            KeyKeeper saved = Loudly.getContext().getKeyKeeper(i);
```
Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java]: Do we really need Loudly.getContext().getKeyKeeper(i) in new architecture?

File: app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java (hunk 5 above)
Comment (danilkolikov) [inline app/src/androidTest/java/ly/loud/loudly/util/database/DatabaseTest.java]: I thought Mockito is for UI, not background

File: app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    @Test
+    public void testDeleteKey() throws DatabaseException {
+        List<KeyKeeper> keys = initRandomKeys();
+        DatabaseUtils.saveKeys();
+
+        DatabaseUtils.deleteKey(Networks.VK);
+        keys.set(Networks.VK, null);
+        Loudly.getContext().setKeyKeeper(Networks.VK, null);
```
Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java]: Again seems like we could use VKModel for this

File: app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java (hunk 6 above)
Comment (danilkolikov) [inline app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java]: Now it works like it. Anyway, now we often use Loudly.getContext(). Until we get rid of it, I think it's acceptable to use it for keys

File: app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java (hunk 7 above)
Comment (danilkolikov) [inline app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java]: And inject model into tests?

File: app/src/main/java/ly/loud/loudly/util/Equality.java [hunk 8]
```diff
@@ -44,21 +41,23 @@
      * Null-safe equality of strings
      * @return true, of strings are equal
      */
-    public static boolean equal(String a, String b) {
-        return equalBuilder(a, b, () -> a.equals(b));
-    }
-
-    public static boolean equal(Uri a, Uri b) {
-        return equalBuilder(a, b, () -> a.equals(b));
+    public static boolean equal(@Nullable String a, @Nullable String b) {
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/util/Equality.java]: What is this function for?

File: app/src/main/java/ly/loud/loudly/util/database/DatabaseUtils.java [hunk 9]
```diff
@@ -102,11 +102,23 @@ static void deleteLocation(long id) throws DatabaseException {
         }
     }
 
-    static <T extends Attachment & MultipleNetwork> long saveAttachment(@NonNull T attachment, long postId) throws DatabaseException {
+    @NonNull
+    private static Link[] instancesToLinks(MultipleNetworkElement element) {
+        Link[] links = new Link[Networks.NETWORK_COUNT];
+        @SuppressWarnings("unchecked") // Java doesn't infer that it's an array of SingleNetworkElement
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/util/database/DatabaseUtils.java]: Emm, why?

File: app/src/main/java/ly/loud/loudly/util/Equality.java (hunk 8 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/util/Equality.java]: Null-safe equality. Like Objects.equal(), but for Java 1.6

File: app/src/main/java/ly/loud/loudly/util/database/DatabaseUtils.java (hunk 9 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/util/database/DatabaseUtils.java]: I don't know, it erases ArrayList<T extends SingleNetworkElement> to ArrayList<Object>, ignoring constraint

Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/util/database/DatabaseUtils.java]: Markdown dropped brackets. There were ArrayList ((T extends SingleNetworkElement)) and ArrayList ((Object))

File: app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java (hunk 7 above)
Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java]: We could mock such models actually

Comment (ZeRoGerc) [inline app/src/androidTest/java/ly/loud/loudly/util/database/KeyTest.java]: when(VKModel.getKeyKeeper).thenReturn(null)

**Comment by ZeRoGerc** [comment]
ok

**Pull request merged**